### PR TITLE
screencopy: Extend protocol to allow linux-dmabuf

### PR DESCRIPTION
**Pull request by any1**
Protocol changes for https://github.com/swaywm/wlroots/pull/2133

Closes https://github.com/swaywm/wlr-protocols/issues/73

### CODE REVIEW

File: unstable/wlr-screencopy-unstable-v1.xml [hunk 1]
```diff
@@ -102,6 +102,12 @@
 
         The client should then create a buffer with the provided attributes, and
         send a "copy" request.
+
+        When the buffer is a dmabuf, the stride parameter may be ignored.
+
+        The format parameter shall be interpreted according to
```
Comment (emersion) [inline unstable/wlr-screencopy-unstable-v1.xml:108]: This part should be covered by https://github.com/swaywm/wlr-protocols/pull/81

**Comment by emersion** [comment]
From a protocol point of view, if we didn't have to maintain backwards compat, I think it would be better to have one event per supported buffer type, each with buffer parameters, followed by a "done" event (which would mark the end of buffer param events).

This would allow users to know exactly which types of buffers are supported. This would avoid having to shoehorn all buffer types params into a single event: `wl_shm.format` shouldn't be used with linux-dmabuf and `stride` may be ignored for linux-dmabuf.

**Comment by any1** [comment]
Updated according to review.

**Comment by emersion** [comment]
I'd prefer having one event per buffer type advertising the buffer parameters.

**Comment by any1** [comment]
> I'd prefer having one event per buffer type advertising the buffer parameters.

Ok, committed something to that effect just now.

File: unstable/wlr-screencopy-unstable-v1.xml [hunk 2]
```diff
@@ -203,5 +203,18 @@
       <arg name="width" type="uint" summary="current width"/>
       <arg name="height" type="uint" summary="current height"/>
     </event>
+
+    <!-- Version 3 additions -->
+    <event name="linux_dmabuf" since="3">
+      <description summary="linux-dmabuf buffer information">
+        Provides information about the frame's linux-dmabuf buffer.
+
+        This event is sent once before the regular "buffer" event if
```
Comment (emersion) [inline unstable/wlr-screencopy-unstable-v1.xml:212]: Maybe we should add a `buffer_done` event instead relying on `buffer`?

**Comment by emersion** [comment]
Descriptions of other requests mention the `buffer` event. They need to be updated.

File: unstable/wlr-screencopy-unstable-v1.xml (hunk 2 above)
Comment (any1) [reply unstable/wlr-screencopy-unstable-v1.xml:212]: Yes, that would make it possible for compositors which implement the protocol to omit `wl_shm`.

**Comment by any1** [comment]
Updated.

**Comment by emersion** [comment]
The new events look good. Please update the `zwlr_screencopy_frame_v1` and `copy` descriptions.

**Comment by any1** [comment]
> The new events look good. Please update the `zwlr_screencopy_frame_v1` and `copy` descriptions.

Done.

File: unstable/wlr-screencopy-unstable-v1.xml [hunk 3]
```diff
@@ -80,13 +80,15 @@
...
       This object represents a single frame.
 
-      When created, a "buffer" event will be sent. The client will then be able
-      to send a "copy" request. If the capture is successful, the compositor
-      will send a "flags" followed by a "ready" event.
+      When created, a series of buffer events will be sent, each representing a
+      supported buffer type. The "buffer_done" event is sent afterwards to
+      indicate that all supported buffer types have been enumerated. The client
+      will then be able to send a "copy" request. If the capture is successful,
+      the compositor will send a "flags" followed by a "ready" event.
```
Comment (emersion) [inline unstable/wlr-screencopy-unstable-v1.xml:91]: Ah, can we just add a clarification here: "For objects version 2 or lower, wl_shm buffers are always supported, ie. the "buffer" event is guaranteed to be sent."

Comment (any1) [reply unstable/wlr-screencopy-unstable-v1.xml:91]: Done

**Comment by emersion** [comment]
Thanks!

**Pull request merged**